### PR TITLE
enable multitouch

### PR DESCRIPTION
**Pull request by p-an**
Hi,
this is reworked #1103. Could you, please, merge it?

Thanks 

Pavel

### CODE REVIEW

**Comment by tuxflo** [comment]
Hi @p-an your commit looks promising to since I'm waiting for multitouch support since I first tried anbox at all. Since you seem to have the necessary SDL knowledge: would it be possible to add stylus support the same way? AFAIK the events are the same just with different IDs.

**Comment by p-an** [comment]
Hi @tuxflo, I am sorry to disappoint you. I don`t have any great knowledge of SDL. All I needed I found in the online documentation. I suppose that adding stylus support is not difficult, but I don`t have any stylus device, so any attempt would be just a blind shot.

**Comment by tuxflo** [comment]
Ok so I'll do some research and then I might try to implement it on my own. BTW I built a Version with your changes and its working great so far. Thanks a lot for implementing this! :+1:

**Comment by ZZZXXX0110** [comment]
I do have a stylus device! I can test for you!

**Comment by tuxflo** [comment]
I also have stylus devices. However SDL doesn't seem to be able to capture the input events from those. I tried to build a non anbox related prototype but SDL only captures mouse events and doesn't even set the SDL_TOUCH_MOUSEID flag. Also calling `SDL_GetNumTouchDevices()` returns just 1, but xinput lists more than one device. It might be easier to directly pass the device into the container but the architecture of anbox does not allow this as far as I can see.

**Comment by JasMich** [comment]
@morphis Are there any plans to merge?

**Comment by Disctanger** [comment]
I am also waiting for merge of this branch.
Does i have built @p-an 's PR but i wonder what should i do to update already installed anbox without losing existing android data.

File: src/anbox/platform/sdl/platform.cpp [hunk 1]
```diff
@@ -125,9 +122,11 @@ Platform::Platform(
   touch_->set_abs_bit(ABS_MT_POSITION_Y);
   touch_->set_abs_max(ABS_MT_POSITION_Y, display_frame.height());
   touch_->set_abs_bit(ABS_MT_TRACKING_ID);
-  touch_->set_abs_max(ABS_MT_TRACKING_ID, 10);
+  touch_->set_abs_max(ABS_MT_TRACKING_ID, MAX_TRACKING_ID);
   touch_->set_prop_bit(INPUT_PROP_DIRECT);
 
+  for(int i = 0; i < MAX_FINGERS ; i++)touch_slots[i] = -1;
```
Comment (morphis) [inline src/anbox/platform/sdl/platform.cpp]: This should be 
```
for (...)
  touch_slots[i] = -1;
```

File: src/anbox/platform/sdl/platform.cpp [hunk 2]
```diff
@@ -204,30 +203,14 @@ void Platform::process_input_event(const SDL_Event &event) {
...
     case SDL_MOUSEBUTTONUP:
       if (config_.no_touch_emulation) {
         mouse_events.push_back({EV_KEY, BTN_LEFT, 0});
       } else {
-        touch_events.push_back({EV_ABS, ABS_MT_TRACKING_ID, -1});
-        touch_events.push_back({EV_ABS, ABS_MT_SLOT, 0});
-        touch_events.push_back({EV_KEY, BTN_TOUCH, 0});
-        touch_events.push_back({EV_KEY, BTN_TOOL_FINGER, 0});
-        touch_events.push_back({EV_SYN, SYN_REPORT, 0});
+        push_finger_up(emulated_touch_id_,touch_events);
```
Comment (morphis) [inline src/anbox/platform/sdl/platform.cpp]: Missing space between `,` and `touch_events`

File: src/anbox/platform/sdl/platform.cpp [hunk 3]
```diff
@@ -248,14 +231,7 @@ void Platform::process_input_event(const SDL_Event &event) {
...
       } else {
-        touch_events.push_back({EV_ABS, ABS_MT_SLOT, 0});
-
-        touch_events.push_back({EV_ABS, ABS_MT_POSITION_X, x});
-        touch_events.push_back({EV_ABS, ABS_MT_POSITION_Y, y});
-
-        touch_events.push_back({EV_ABS, ABS_MT_TOUCH_MAJOR, 24});
-        touch_events.push_back({EV_ABS, ABS_MT_TOUCH_MINOR, 24});
-        touch_events.push_back({EV_SYN, SYN_REPORT, 0});
+        push_finger_motion(x,y,emulated_touch_id_,touch_events);
```
Comment (morphis) [inline src/anbox/platform/sdl/platform.cpp]: See comment above

File: src/anbox/platform/sdl/platform.cpp [hunk 4]
```diff
@@ -285,44 +261,21 @@ void Platform::process_input_event(const SDL_Event &event) {
...
     // Touch screen
     case SDL_FINGERDOWN: {
-      touch_events.push_back({EV_ABS, ABS_MT_TRACKING_ID, static_cast<std::int32_t>(event.tfinger.fingerId)});
-      touch_events.push_back({EV_ABS, ABS_MT_SLOT, 0});
-      touch_events.push_back({EV_KEY, BTN_TOUCH, 1});
-      touch_events.push_back({EV_KEY, BTN_TOOL_FINGER, 1});
-
       if (!calculate_touch_coordinates(event, x, y))
         break;
+      push_finger_down(x,y,event.tfinger.fingerId,touch_events);
```
Comment (morphis) [inline src/anbox/platform/sdl/platform.cpp]: See comment above

File: src/anbox/platform/sdl/platform.cpp [hunk 5]
```diff
@@ -285,44 +261,21 @@ void Platform::process_input_event(const SDL_Event &event) {
...
       break;
     }
-	  case SDL_FINGERUP: {
-      touch_events.push_back({EV_ABS, ABS_MT_TRACKING_ID, -1});
-      touch_events.push_back({EV_ABS, ABS_MT_SLOT, 0});
-      touch_events.push_back({EV_KEY, BTN_TOUCH, 0});
-      touch_events.push_back({EV_KEY, BTN_TOOL_FINGER, 0});
-      touch_events.push_back({EV_SYN, SYN_REPORT, 0});
+    case SDL_FINGERUP: {
+      push_finger_up(event.tfinger.fingerId,touch_events);
```
Comment (morphis) [inline src/anbox/platform/sdl/platform.cpp]: See comment above

File: src/anbox/platform/sdl/platform.cpp [hunk 6]
```diff
@@ -341,6 +294,53 @@ void Platform::process_input_event(const SDL_Event &event) {
...
+    if(last_slot != slot){
+        touch_events.push_back({EV_ABS, ABS_MT_SLOT, slot});
+        last_slot = slot;
+    }
+}
+
+void Platform::push_finger_down(int x, int y, int finger_id, std::vector<input::Event> &touch_events){
+    int slot = find_touch_slot(-1);
+    if(slot == -1){
+        DEBUG("no free slot!\n");
```
Comment (morphis) [inline src/anbox/platform/sdl/platform.cpp]: There is no `\n` necessary

Comment (p-an) [reply src/anbox/platform/sdl/platform.cpp]: Fixed. I hope I didn`t miss anything. I am sorry for the delay.

P.

**Comment by morphis** [comment]
@p-an The code still isn't following the Anbox code style. Can you fix that? Once that is done I am happy to merge this

**Comment by p-an** [comment]
Hi,
I fixed some more. If I missed something else, could you, please, point me to specific lines?

Thanks

Pavel

File: src/anbox/platform/sdl/platform.cpp [hunk 7]
```diff
@@ -341,6 +295,53 @@ void Platform::process_input_event(const SDL_Event &event) {
     touch_->send_events(touch_events);
 }
 
+int Platform::find_touch_slot(int id){
+    for(int i = 0; i < MAX_FINGERS; i++){
+        if(touch_slots[i] == id)return i;
+    }
```
Comment (morphis) [inline src/anbox/platform/sdl/platform.cpp]: This should be 
```
for (int i = 0; ...) {
  if (touch_slots[i] == id)
    return i;
}
```

File: src/anbox/platform/sdl/platform.cpp [hunk 8]
```diff
@@ -341,6 +295,53 @@ void Platform::process_input_event(const SDL_Event &event) {
...
+    return -1;
+}
+
+void Platform::push_slot(std::vector<input::Event> &touch_events, int slot){
+    if(last_slot != slot){
```
Comment (morphis) [inline src/anbox/platform/sdl/platform.cpp]: This misses spaces between `if` and `(` and `)` and `{`

File: src/anbox/platform/sdl/platform.cpp [hunk 9]
```diff
@@ -341,6 +295,53 @@ void Platform::process_input_event(const SDL_Event &event) {
...
+        touch_events.push_back({EV_ABS, ABS_MT_SLOT, slot});
+        last_slot = slot;
+    }
+}
+
+void Platform::push_finger_down(int x, int y, int finger_id, std::vector<input::Event> &touch_events){
+    int slot = find_touch_slot(-1);
+    if(slot == -1){
```
Comment (morphis) [inline src/anbox/platform/sdl/platform.cpp]: See above

File: src/anbox/platform/sdl/platform.cpp [hunk 10]
```diff
@@ -341,6 +295,53 @@ void Platform::process_input_event(const SDL_Event &event) {
...
+    push_slot(touch_events, slot);
+    touch_events.push_back({EV_ABS, ABS_MT_TRACKING_ID, static_cast<std::int32_t>(finger_id % MAX_TRACKING_ID + 1)});
+    touch_events.push_back({EV_ABS, ABS_MT_POSITION_X, x});
+    touch_events.push_back({EV_ABS, ABS_MT_POSITION_Y, y});
+    touch_events.push_back({EV_SYN, SYN_REPORT, 0});
+}
+
+void Platform::push_finger_up(int finger_id, std::vector<input::Event> &touch_events){
+    int slot = find_touch_slot(finger_id);
+    if(slot == -1)return;
```
Comment (morphis) [inline src/anbox/platform/sdl/platform.cpp]: See above, return belongs in a new line

File: src/anbox/platform/sdl/platform.cpp [hunk 11]
```diff
@@ -341,6 +295,53 @@ void Platform::process_input_event(const SDL_Event &event) {
...
+    push_slot(touch_events, slot);
+    touch_events.push_back({EV_ABS, ABS_MT_TRACKING_ID, -1});
+    touch_events.push_back({EV_SYN, SYN_REPORT, 0});
+    touch_slots[slot] = -1;
+}
+
+void Platform::push_finger_motion(int x, int y, int finger_id, std::vector<input::Event> &touch_events){
+    int slot = find_touch_slot(finger_id);
+    if(slot == -1)return;
```
Comment (morphis) [inline src/anbox/platform/sdl/platform.cpp]: See above

File: src/anbox/platform/sdl/platform.h [hunk 12]
```diff
@@ -97,6 +98,16 @@ class Platform : public std::enable_shared_from_this<Platform>,
   bool window_size_immutable_ = false;
   std::uint32_t focused_sdl_window_id_ = 0;
   Configuration config_;
+
+  enum{MAX_FINGERS = 10};
+  enum{MAX_TRACKING_ID = 10};
```
Comment (morphis) [inline src/anbox/platform/sdl/platform.h]: These are most constants rather than enums.

File: src/anbox/platform/sdl/platform.h [hunk 13]
```diff
@@ -97,6 +98,16 @@ class Platform : public std::enable_shared_from_this<Platform>,
...
+  int touch_slots[MAX_FINGERS];
+  int last_slot = -1;
+  int find_touch_slot(int id);
```
Comment (morphis) [inline src/anbox/platform/sdl/platform.h]: Methods and member variables have to be separated with an empty line in between,

**Comment by morphis** [comment]
@p-an Added comments for the remaining ones. Thanks for fixing!

**Pull request merged**